### PR TITLE
Fix get csr memory leak

### PR DESCRIPTION
**Pull request by Wenxing-hou**
Fix: #932 
Fix: #933 

Signed-off-by: Wenxing Hou <wenxing.hou@intel.com>

### CODE REVIEW

**Comment by jyao1** [comment]
I think below function should be changed:
```
bool libspdm_gen_csr(uint32_t base_hash_algo, uint32_t base_asym_algo, bool *need_reset,
                     size_t *csr_len, uint8_t **csr_pointer, size_t csr_buffer_size,
                     uint8_t *requester_info, size_t requester_info_length);
```

to
```
bool libspdm_gen_csr(uint32_t base_hash_algo, uint32_t base_asym_algo, bool *need_reset,
                     uint8_t *requester_info, size_t requester_info_length,
                     uint8_t *csr, size_t *csr_len);
```

for input csr_len is csr buffer max size.
for output csr_len is csr buffer actual size.

File: os_stub/spdm_device_secret_lib_sample/lib.c [hunk 1]
```diff
@@ -266,21 +266,33 @@ bool libspdm_gen_csr(uint32_t base_hash_algo, uint32_t base_asym_algo, bool *nee
...
-        /*get the cahed requester info and csr*/
+        /*get the cached requester info and csr*/
         if ((result) &&
             (cached_req_info_length == requester_info_length) &&
             (libspdm_const_compare_mem(cached_req_info, requester_info, requester_info_length)) &&
-            (libspdm_read_cached_csr(base_asym_algo, csr_pointer, csr_len))) {
+            (libspdm_read_cached_csr(base_asym_algo, &cached_csr, csr_len))) {
+
+            /*get and save cached csr*/
+            libspdm_copy_mem(*csr_pointer, csr_buffer_size, cached_csr, *csr_len);
```
Comment (jyao1) [inline os_stub/spdm_device_secret_lib_sample/lib.c:283]: You cannot do the copy directly.

You can compare the size and return BUFFER_TOO_SMALL, if the destination is not big enough.

Comment (Wenxing-hou) [reply os_stub/spdm_device_secret_lib_sample/lib.c]: Yes. I have added the check and deleted useless parameter :csr_buffer_size.

File: include/library/spdm_device_secret_lib.h [hunk 2]
```diff
@@ -234,17 +234,17 @@ bool libspdm_write_certificate_to_nvm(uint8_t slot_id, const void * cert_chain,
...
+ *                                       For input, csr_pointer is address for stored CSR.
+ *
  * @retval  true   Success.
  * @retval  false  Failed to gen CSR.
  **/
 bool libspdm_gen_csr(uint32_t base_hash_algo, uint32_t base_asym_algo, bool *need_reset,
-                     size_t *csr_len, uint8_t **csr_pointer, size_t csr_buffer_size,
-                     uint8_t *requester_info, size_t requester_info_length);
+                     uint8_t *requester_info, size_t requester_info_length,
+                     size_t *csr_len, uint8_t **csr_pointer);
```
Comment (jyao1) [inline include/library/spdm_device_secret_lib.h:250]: Why you need **csr_pointer? Why not use *csr, as I suggested?

Comment (jyao1) [reply include/library/spdm_device_secret_lib.h:250]: OK, then please describe that clearly in the comment

Comment (Wenxing-hou) [reply include/library/spdm_device_secret_lib.h:251]: Yes. I have added the comment. Thanks.

**Review by steven-bellock — APPROVED** [review]

**Pull request merged**